dali/ft/filecopy.cpp

```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
                 if (stricmp(aname, "Protect") != 0)
                     curProps.addPropTree(aname, createPTreeFromIPT(&iter->query()));
             }
+
+            // Handle history
+            IPropertyTree *curHistory = lock.queryHistory();
+
+            if (srcHistory)
+            {
+                // copy history records from source file
+                Owned<IPropertyTreeIterator> historyIter = srcHistory->getElements("*");
+                ForEach(*historyIter)
+                {
+                    const char *aname = historyIter->query().queryName();
+                    curHistory->addPropTree(aname, createPTreeFromIPT(&historyIter->query()));
+                }
+            }
+
+            // Add new record about this operation
+            IPropertyTree * newRecord = createPTree("Origin");
+
+            CDateTime temp;
+            temp.setNow();
+            unsigned hour, min, sec, nanosec;
+            temp.getTime(hour, min, sec, nanosec);
+            temp.setTime(hour, min, sec, 0);
+
+            StringBuffer timestr;
+            newRecord->addProp("@timestamp",temp.getString(timestr).str());
+
+            StringBuffer val;
+            srcAttr->getProp("@owner", val);
+            newRecord->addProp("@owner", val.str());
+
+            srcAttr->getProp("@workunit", val.clear());
+            if (0 == val.length())
+                val.set("N/A");
+            newRecord->addProp("@workunit", val.str());
+
+            newRecord->addProp("@operation", getOperationTypeString());
+
+            // add original file name
+            if (distributedSource)
+            {
+                RemoteFilename remoteFile;
+                distributedSource->getPart(0)->getFilename(remoteFile, 0 );
+                splitAndStoreFileInfo(newRecord, remoteFile);
+            }
+            else
+            {
+                val.clear();
+                ForEachItemIn(idx, sources)
+                {
+                    FilePartInfo & curSource = sources.item(idx);
+                    RemoteFilename remoteFile = curSource.filename;
+                    splitAndStoreFileInfo(newRecord, remoteFile, idx, false);
```

Comment:
Originally that code was two places in true and false branch (as you mentioned in this comment) but I heard in my mind you comment about similar code should summed. 
And the other reason to create this function was: both code worked on a RemoteFileName object and doing same separation and assignments.